subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +460,399 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
 
 	return ret;
 }
+
+#if defined(CONFIG_TLS_CREDENTIAL_KEYGEN)
+
+#include <psa/crypto.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/asn1.h>
+
+/* For this backend, the private-key will be stored in DER format.
+ * Use `cred get <tag> PK bin` to retrieve (the non-terminated base64 encoding of) the key using
+ * the credential shell.
+ */
+int tls_credential_keygen(sec_tag_t tag, enum tls_credential_keygen_type type,
+			  void *key_buf, size_t *key_len)
+{
+	int ret = 0;
+
+	/* PSA keygen */
+	psa_status_t psa_status;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_usage_t key_usage;
+
+	/* MbedTLS key formatting */
+	int mbed_status;
+	mbedtls_pk_context pk_ctx;
+	void* key_start = 0;
+	size_t key_max;
+
+	/* Permanent key storage */
+	int cred_status;
+	psa_storage_uid_t uid = tls_credential_get_uid(tag, TLS_CREDENTIAL_PRIVATE_KEY);
+	bool key_stored = false;
+
+	if (key_len == NULL) {
+		return -EINVAL;
+	}
+
+	key_max = *key_len;
+
+	k_mutex_lock(&credential_lock, K_FOREVER);
+	mbedtls_pk_init(&pk_ctx);
+
+	/* Ensure the storage destination is valid. */
+	if (!tag_type_valid(tag, TLS_CREDENTIAL_PRIVATE_KEY)) {
+		ret = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Presently, only SECP256R1 is supported by this backend for keygen */
+	if (!tls_credential_can_keygen(type)) {
+		ret = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Verify sectag not already taken */
+	if (tls_credential_toc_find_slot(uid) != CRED_MAX_SLOTS) {
+		ret = -EEXIST;
+		goto cleanup;
+	}
+
+	/* Before attempting keygen, check that an empty slot is available */
+	if (tls_credential_toc_find_slot(0) == CRED_MAX_SLOTS) {
+		ret = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Use PSA to generate a volatile SECP256R1 private/public key-pair
+         *
+	 * It is possible (and better) to mark the private key as persistent, and store it directly
+	 * in Trusted Internal Storage. This would be more secure, but until we support opaque
+	 * keys properly, it isn't an option.
+	 *
+	 * Accordingly, for now, we simply generate a volatile key and transfer it into
+	 * protected storage.
+	 */
+
+	key_usage = PSA_KEY_USAGE_EXPORT;
+	psa_set_key_usage_flags(&key_attributes, key_usage);
+
+	psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+	psa_set_key_algorithm(&key_attributes, PSA_ALG_ECDSA(PSA_ALG_SHA_256));
+	psa_set_key_type(&key_attributes, PSA_KEY_TYPE_ECC_KEY_PAIR(PSA_ECC_FAMILY_SECP_R1));
+	psa_set_key_bits(&key_attributes, 256);
+
+	psa_status = psa_crypto_init();
```

Comment:
Since you are also including Mbed TLS in the build, this call is already done by Mbed TLS[ initialization](https://github.com/zephyrproject-rtos/zephyr/blob/52540d8412630edb2c64f4fcca1c1165bcce1ac3/modules/mbedtls/zephyr_init.c#L57) so  I think this can be removed.

Reply:
You're absolutely right that it probably can in practice be removed, but in principle I think I am not supposed to assume that when interacting directly with PSA crypto APIS like this. For example if someone manages to get this running on a non-TFM build. (though I am not actually sure whether that is possible, tbh)

 Is there any downside to keeping the call?